File: homework/Van_Le/homework4/RandomizedSet.cs
namespace running_app
{
    public class RandomizedSet
    {
        private Dictionary<int, int> _dict;
        private List<int> _list;
        private Random _rand;

        // Initialize data structure
        public RandomizedSet()
        {
            _dict = new Dictionary<int, int>();
            _list = new List<int>();
            _rand = new Random();
        }

        // Inserts a value to the set. Returns true if the set did not already contain the specified element.
        public bool Insert(int val)
        {
            if (_dict.ContainsKey(val))
                return false;
            _list.Add(val);
            int index = _list.Count - 1;
            _dict[val] = index;
            return true;
        }

        // Removes a value from the set. Returns true if the set contained the specified element.
        public bool Remove(int val)
        {
            if (!_dict.ContainsKey(val))
                return false;
            int index = _dict[val];

            // to pop at O(1) complexity time, it needs to swap the desired position and pop that out
            (_list[index], _list[^1]) = (_list[^1], _list[index]);

            _dict[_list[index]] = index;
            _list.RemoveAt(_list.Count - 1);
            _dict.Remove(val);
            return true;
        }

        // Get a random element from the set. 
        public int GetRandom()
        {
            int index = _rand.Next(0, _list.Count);
            return _list[index];
        }
    }
}

File: homework/Long_Nguyen/lesson_4/RandomizedSet.java
import java.util.*;

class RandomizedSet {
    private Map<Integer, Integer> map; // key: val, value: index
    private List<Integer> list;
    private Random rand = new Random();

    public RandomizedSet() {
        this.map = new HashMap<>();
        this.list = new ArrayList<>();
    }

    public boolean insert(int val) {
        if (map.containsKey(val)) {
            return false;
        }
        list.add(val);
        map.put(val, list.size() - 1);

        return true;
    }

    public boolean remove(int val) {
        if (!map.containsKey(val)) {
            return false;
        }
        int removeElementIdx = map.get(val);
        int lastElementIdx = list.size() - 1;
        int lastElementVal = list.get(lastElementIdx);

        int temp = list.get(removeElementIdx);
        list.set(removeElementIdx, lastElementVal);
        list.set(lastElementIdx, temp);

        list.remove(lastElementIdx);
        map.put(lastElementVal, removeElementIdx);
        map.remove(val);

        return true;
    }

    public int getRandom() {
        return list.get(rand.nextInt(list.size()));
    }
}
